### PR TITLE
Particles-in-Cell, multi-phase rheologies & others

### PR DESCRIPTION
**Pull request by albert-de-montserrat**
Particles-in-Cell:
- [x] Coupling with [JustPIC.jl](https://github.com/JuliaGeodynamics/JustPIC.jl)
- [x] Multi-phase methods based on PiC - heavylifting done with [CellArrays.jl](https://github.com/omlins/CellArrays.jl)
- [x] Stokes and heat diffusion solvers now work with multiple rheological phases, based on particles.
- [x] Stress rotation on particles

Mini-apps:
- [x] [Van Keken](https://agupubs.onlinelibrary.wiley.com/doi/pdf/10.1029/97JB01353) thermochemical [benchmark](https://github.com/PTsolvers/JustRelax.jl/blob/1e5e9a47f51c4161c6b6aac7447a45a6abf0fc16/miniapps/benchmarks/stokes2D/VanKeken.jl/VanKeken.jl#L1) with particles-in-cell
- [x] Multi-phase thermo-mechanical convection with PiC
- [x] Heat diffusion [benchmark](https://github.com/PTsolvers/JustRelax.jl/blob/1e5e9a47f51c4161c6b6aac7447a45a6abf0fc16/miniapps/benchmarks/thermal_diffusion/diffusion/diffusion2D_multiphase.jl#L1) using the Pseudo-Transient method with multiple phases

Others:
- [x] Improved convergence of Stokes solver for non-Newtonian rheologies.
- [x] Corrections on interpolations from/to different cell locations.
- [x] Replace dimension-dependant ParalleStencil macros by closures and [functions](https://github.com/PTsolvers/JustRelax.jl/blob/adm-justpic/src/MiniKernels.jl)
- [x] `compute_maxloc!` with user defined [window](https://github.com/PTsolvers/JustRelax.jl/blob/1e5e9a47f51c4161c6b6aac7447a45a6abf0fc16/src/Utils.jl#L292).
- [x] Re-organization of `stokes2d.jl`; Stokes-related kernels moved into their own files.
- [ ] ~~Re-organization of `stokes3d.jl`, as in the 2D version~~ (for next PR)
- [x] Collection of bugfixes.

### CODE REVIEW

File: mwe.jl
Comment (aelligp) [inline mwe.jl:1]: Do you want to have this in the PR?

File: src/rheology/Viscosity.jl [hunk 1]
```diff
@@ -30,32 +21,109 @@ end
...
+        # compute and update stress viscosity
+        ηi = JustRelax.compute_phase_viscosity_εII(rheology, ratio_ij, εII, args_ij)
+        ηi = JustRelax.continuation_log(ηi, η[I...], ν)
+        η[I...] = clamp(ηi, cutoff...)
+    end
+
+    return nothing
+end
+
+function compute_viscosity!(η::CuArray{T1,N,T2}, x...) where {T1,T2,N}
```
Comment (aelligp) [inline src/rheology/Viscosity.jl:76]: testing or did you mean to add this with CuArray specifically?

File: mwe.jl
Comment (albert-de-montserrat) [reply mwe.jl:1]: deleted

File: src/rheology/Viscosity.jl (hunk 1 above)
Comment (albert-de-montserrat) [reply src/rheology/Viscosity.jl:76]: Was testing, and indeed this change should be reversed

**Pull request merged**